jib-core/src/main/java/com/google/cloud/tools/jib/ncache/DefaultCacheStorageWriter.java

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright 2018 Google LLC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.ncache;
+
+import com.google.cloud.tools.jib.blob.Blob;
+import com.google.cloud.tools.jib.blob.BlobDescriptor;
+import com.google.cloud.tools.jib.blob.Blobs;
+import com.google.cloud.tools.jib.filesystem.FileOperations;
+import com.google.cloud.tools.jib.filesystem.TemporaryDirectory;
+import com.google.cloud.tools.jib.hash.CountingDigestOutputStream;
+import com.google.cloud.tools.jib.image.DescriptorDigest;
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.nio.file.AtomicMoveNotSupportedException;
+import java.nio.file.FileAlreadyExistsException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.zip.GZIPOutputStream;
+
+/** Writes to the default cache storage engine. */
+class DefaultCacheStorageWriter {
+
+  /** Holds information about a layer that was written. */
+  private static class WrittenLayer {
+
+    private final DescriptorDigest layerDigest;
+    private final DescriptorDigest layerDiffId;
+    private final long layerSize;
+
+    private WrittenLayer(
+        DescriptorDigest layerDigest, DescriptorDigest layerDiffId, long layerSize) {
+      this.layerDigest = layerDigest;
+      this.layerDiffId = layerDiffId;
+      this.layerSize = layerSize;
+    }
+  }
+
+  /**
+   * Attempts to move {@code source} to {@code destination}. If {@code destination} already exists,
```

Comment:
we talked offline about this, but just making note that we need to overwrite selector files.